dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -0,0 +1,264 @@
+        // maybeExecute may take sufficiently long that other requests could be queued.
+        if (queueSizeEstimate.get() >= maxQueueSize) {
+            return Optional.empty();
+        }
+
+        DeferredCall components = DeferredCall.builder()
+                .endpoint(endpoint)
+                .request(request)
+                .response(SettableFuture.create())
+                .span(DetachedSpan.start("Dialogue-request-enqueued"))
+                .build();
+
+        if (!queuedCalls.offer(components)) {
+            // Should never happen, ConcurrentLinkedDeque has no maximum size
+            return Optional.empty();
+        }
+        queueSizeEstimate.incrementAndGet();
+
+        schedule();
+
+        return Optional.of(components.response());
+    }
+
+    private void onCompletion() {
+        numRunningRequests.decrementAndGet();
+        schedule();
+    }
+
+    /**
+     * Try to schedule as many tasks as possible. Called when requests are submitted and when they complete.
+     */
+    void schedule() {
+        int numScheduled = 0;
+        while (scheduleNextTask()) {
+            numScheduled++;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Scheduled {} requests", SafeArg.of("numScheduled", numScheduled));
+        }
+    }
+
+    /**
+     * Get the next call and attempt to execute it. If it is runnable, wire up the underlying future to the one
+     * previously returned to the caller. If it is not runnable, add it back into the queue. Returns true if more
+     * tasks may be able to be scheduled, and false otherwise.
+     */
+    private boolean scheduleNextTask() {
+        DeferredCall queueHead = queuedCalls.poll();
+        if (queueHead == null) {
+            return false;
+        }
+        SettableFuture<Response> queuedResponse = queueHead.response();
+        // If the future has been completed (most likely via cancel) the call should not be queued.
+        // There's a race where cancel may be invoked between this check and execution, but the scheduled
+        // request will be quickly cancelled in that case.
+        if (queuedResponse.isDone()) {
+            queueSizeEstimate.decrementAndGet();
+            return true;
+        }
+        try (CloseableSpan ignored = queueHead.span().childSpan("Dialogue-request-scheduled")) {
+            Endpoint endpoint = queueHead.endpoint();
+            Optional<ListenableFuture<Response>> maybeResponse = delegate.maybeExecute(endpoint, queueHead.request());
+
+            if (maybeResponse.isPresent()) {
+                queueSizeEstimate.decrementAndGet();
+                ListenableFuture<Response> response = maybeResponse.get();
+                queueHead.span().complete();
+                numRunningRequests.incrementAndGet();
+                response.addListener(numRunningRequests::decrementAndGet, DIRECT);
+                Futures.addCallback(response, new ForwardAndSchedule(queuedResponse), DIRECT);
```

Comment:
```suggestion
                DialogueFutures.addDirectCallback(response, new ForwardAndSchedule(queuedResponse));
```

Reply:
This reintroduces the old queue, I’d prefer to make improvements separately

Reply:
Oh it didn't seem like this was just a raw revert

Reply:
The only diff is that this doesn’t include the two broken gauges